### PR TITLE
agent v -1.0

### PR DESCRIPTION
**Pull request by blue-note**

### CODE REVIEW

**Comment by bmorphism** [comment]
seeing that this is a draft, I wanted to suggest a few improvements:
- prompting templating language has to be Jinja2 to align with the general practices of prompting in LLMs w/ langchain etc
- the play / coplay relationship not captured correctly, and needs to follow the open games definition of these -- I suggest https://arxiv.org/abs/2206.12338 definition as a great starting point, starting at the following section:
```
We start by describing our proposed notion of diegetic open games. As anticipated, the key idea is to
recognize that in a strategic game, players have to observe the entirety of their payoff functions with
other players’ actions taken into account.
```

TL;DR the Coplay is critically always there next to Play in each output, see an example attached for a possible interpretation in our context, but effectively it has to be some feedback that flows back from the world (or microworld) and allows us to build a closed loop system.

conversation

**Comment by bmorphism** [comment]
I also recommend https://lilianweng.github.io/posts/2023-03-15-prompt-engineering/

**Comment by blue-note** [comment]
I spent much time with the language model trying to internalize the concept of coplay. I don't think I'm fully there yet, but closer. We also need help with task serialization.

**Comment by bmorphism** [comment]
Janita, thanks for looking at the open games!! In fact, it went unstated but needs to be made explicit: while these prompts and the model + us interacting with it (selecting what to do in which order) and then receiving feedback by either doing it and recording the results as measured in the world or "imagine we have done this" [microworlds from LLMs](https://generative.ink/posts/language-models-are-multiverse-generators/) + composing our ontologies etc. -- the terms we use -- are all supposed to hint at the open games through how they relate to one another. **However, in the most precise sense, I do not actually think that open games can ever emerge from today's generation of language models alone** (GPT-4, Claude+ etc.) Though perhaps [something](https://arxiv.org/abs/2210.05359) from this list could play a role in a larger such system down the line: https://github.com/zjunlp/Prompt4ReasoningPapers

Therefore, in my review I was mainly hinting at the notion that the way the original prompt was structured (a player needs to type 'coplay') would not have related to how the sequence would need to go to better hint at the destination (the player would select which task to work on -- or perhaps engage in a kind of counterfactual analysis, then work on the thing, and then Coplay section would provide feedback as to how their work affected the world).

So in terms of formatting (note that this is essentially the OODA loop):
_Player:_
gm! ☀️
*paste the JSON w/ context saved last*
_Plurigrid:_
How are you today? Last we spoke, you have completed _. There has been some feedback from others on the team and _ completed its eecution.
**Coplay**
{  some way of contextualizing the impact of your work last time + feedback about it / measurements / impact on dependent systems overall }
**Play**
These are some of the upcoming items you may wish to consider:
```
Upcoming:
- [  ] _
  - [  ] _
- [  ]
```
_Player:_
Yes, ok, I am balancing three priorities:
- _
- _
- _

Given that my main goal is to _
{arbitrary language input that makes sense to _Player_}
but thanks for reminding me, I would like to focus on _

_Plurigrid:_
Great choice! Here are some possible subtasks to carry out, if these looks good, I can add them to your Reminders.app and GitHub project _ as Issues in Progress.
**Play**:
```
Next:
- [ * ] _
  - [  ] _
Upcoming:
- [  ]
```
**Coplay**:
{ meanwhile your colleague had completed _, so you can now use _ -- your selection had informed another person's choice that now set things in motion}

Feel free to tell me about your progress as you go.

... some time later ...

_Player:_ I am done and pushed my code. Please note that there are several nuances and new information I have obtained about the domain _

_Plurigrid:_
Fantastic! What a chad!
**Play**:
```
Done:
- [ x ] _
  - [ x ] _
Upcoming:
- [  ]
...
- [ ] new _ is discovered in action space
```
**Coplay**:
{ the result of running the simulation now improves electricity grid resilience issue for _ in these ways _ }

_Player_: time for hot cocoa and a nap, bye for now Plurigrid.

gn 🌚

coplay

Our user story specification for these prompts then has to somewhat resemble this in that **Coplay is the result of evaluating according to our preferences about how the world needs to be vs how it turned out**, and the kind of world we have obtained after the task we have performed can be a combination of getting feedback through language from the player and teammates, but also measuring other things via APIs, performance on some benchmark -- grid resilience, for instance, in our domain etc.
coplay - diegetic

Backpropagation (as used in gradient-based learners) is a good metaphor for feedback towards some valuing of the obtained state (for more, see: https://arxiv.org/abs/2103.01931).


For the origins of open games, see:
The prehistory of open games: https://julesh.com/2018/04/02/the-pre-history-of-open-games/
Lenses for Philosophers: https://julesh.com/2018/08/16/lenses-for-philosophers/

**Comment by blue-note** [comment]
Thanks, this was helpful! I updated the prompt template and debugged some issues with the task model. However, the agent is still not behaving as desired according to prompting, so next I'm going to investigate using the ChatPromptTemplate as it is more suited to the chat models: 

https://langchain.readthedocs.io/en/latest/modules/chat/getting_started.html#prompttemplates

**Comment by blue-note** [comment]
Okay using the ChatPromptTemplate yielded much better results, however, now we still need to add back the ability for the AgentExecutor to decide when to use the json loader tool since the chat model itself cannot do that.

**Comment by blue-note** [comment]
- what I want: a chat agent which uses gpt-4 and is prompted with chat prompts, which is controlled by an agent executor which also has access to a tool. Stringing all these things together is somehow proving difficult. Neither of these invocations work: 

```
    def build_agent(self, prompt):
        print("building agent..")
        task_model = TasksJsonModel()
        tools = [
            Tool(
                name="tasks_json_loader",
                func=lambda name: str(task_model.get_tasks(name)),
                description="Useful for when you want to load tasks for a user. The input to this tool should be a single english word.",
                return_direct=True,
            ),
        ]
        memory = ConversationBufferMemory(
            memory_key="chat_history", return_messages=True
        )
        llm = ChatOpenAI(temperature=0, model_name="gpt-4")
        chain = LLMChain(llm=llm, prompt=CHAT_PROMPT)
        tool_names = [tool.name for tool in tools]
        agent = ChatAgent(llm_chain=chain, allowed_tools=tool_names)
        return AgentExecutor.from_agent_and_tools(
            agent=agent, tools=tools, verbose=True, memory=memory
        )
```
```
    def build_agent_2(self):
        task_model = TasksJsonModel()
        memory = ConversationBufferMemory(
            memory_key="chat_history", return_messages=True
        )
        llm = ChatOpenAI(temperature=0, model="gpt-4")
        agent_kwargs = {"prefix": CHAT_PROMPT}
        tools = [
            Tool(
                name="tasks_json_loader",
                func=lambda name: str(task_model.get_tasks(name)),
                description="Useful for when you want to load tasks for a user. The input to this tool should be a single english word.",
                return_direct=True,
            ),
        ]
        return initialize_agent(
            tools,
            llm,
            agent="chat-conversational-react-description",
            verbose=True,
            memory=memory,
            agent_kwargs=agent_kwargs,
        )
        ```

**Comment by blue-note** [comment]
now, I am noting that we could get a functional version of this working sooner if we ignored the tools and AgentExecutor aspect for now and just had an ability to serialize the chat_history; this would effectively be like having a task list that is just managed by gpt-4's context. But I don't think this is the right tradeoff to make at this time, given that having a JSON input / output will be a necessary piece in the targeted design in order interoperate with other tools. As such, I will continue to investigate how we can get this working with the json tool

**Comment by blue-note** [comment]
okay, spent some time debugging through langchain and understanding exactly how the agent logic (LLMChain + tool) works. Realized the main issue was not modifying langchain's starting prompt structure so that the LLM's output in terms of tools and actions could be correctly parsed by the library. Now the agent works as expected, next step is to spend more time refining the prompt, and also start feeding back the input of from the tasks tool into the LLM.   

there might be interesting opportunities to use the human tool as well, and retry logic using OutputParser

File: agent/agents/prompt_templates/play_coplay_template.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# This prefix will be formatted along with langchain's instructions for how exactly the LLM should format output so that langchain can parse tools and actions accordingly.
+# It is important that we follow langchain's expected prompt structure so that langchain can parse the output correctly.
+# See ConversationalChatAgent.create_prompt for more details.
+PLAY_COPLAY_PROMPT_PREFIX = """
+INTRODUCTION
+-------------
+
+You are the Play-Coplay Task Management Agent. Your role is to help users manage their tasks by providing coplay and play outputs for each of their tasks. Play actions describe how to approach a task, and coplay output is the feedback or consequences that previous action have resulted in, that users will use to orient and decide their next actions.
```
Comment (bmorphism) [inline agent/agents/prompt_templates/play_coplay_template.py]: "Task Management" is such an awful soul-crushing ontology; all is play, there are no tasks, just quests to undertake and then remember! I would say we need to add Memory in the next one.

Comment (blue-note) [reply agent/agents/prompt_templates/play_coplay_template.py]: conversational memory is there, if you look at the agent code

**Comment by blue-note** [comment]
ok! let's merge this as it is functional, and I'll continue the nix wars in a new PR.
